### PR TITLE
[WIP] Pubsub ordering keys

### PR DESCRIPTION
**Pull request by sduskis**

### CODE REVIEW

File: google-cloud-clients/google-cloud-pubsub/src/main/java/com/google/cloud/pubsub/v1/Publisher.java [hunk 1]
```diff
@@ -84,15 +89,17 @@
   private final String topicName;
 
   private final BatchingSettings batchingSettings;
+  private final boolean enableMessageOrdering;
 
   private final Lock messagesBatchLock;
-  private MessagesBatch messagesBatch;
+  private final Map<String, MessagesBatch> messagesBatches;
```
Comment (sduskis) [inline google-cloud-clients/google-cloud-pubsub/src/main/java/com/google/cloud/pubsub/v1/Publisher.java]: Perhaps we should consider the benefits of having a `ConcurrentHashMap`, we'd be able to parallelize requests that way.

File: google-cloud-clients/google-cloud-pubsub/src/main/java/com/google/cloud/pubsub/v1/Publisher.java [hunk 2]
```diff
@@ -351,21 +384,45 @@ public void onFailure(Throwable t) {
               messagesWaiter.incrementPendingMessages(-outstandingBatch.size());
             }
           }
-        },
-        directExecutor());
+        };
+
+    if (outstandingBatch.orderingKey == null || outstandingBatch.orderingKey.isEmpty()) {
```
Comment (sduskis) [inline google-cloud-clients/google-cloud-pubsub/src/main/java/com/google/cloud/pubsub/v1/Publisher.java]: Consider an encapsulating method for `hasOrderingKey()`

File: google-cloud-clients/google-cloud-pubsub/src/main/java/com/google/cloud/pubsub/v1/Publisher.java [hunk 3]
```diff
@@ -128,9 +137,31 @@ private Publisher(Builder builder) throws IOException {
...
-    if (retrySettings.getMaxAttempts() == 0) {
-      retrySettings = retrySettings.toBuilder().setMaxAttempts(Integer.MAX_VALUE).build();
+    // Also, if "message ordering" is enabled, the publisher should retry sending the failed
+    // message infinitely rather than sending the next one.
+    RetrySettings.Builder retrySettingsBuilder = builder.retrySettings.toBuilder();
+    if (retrySettingsBuilder.getMaxAttempts() == 0) {
+      retrySettingsBuilder.setMaxAttempts(Integer.MAX_VALUE);
+    }
+    if (enableMessageOrdering) {
+      retrySettingsBuilder
```
[the remark this answers is not in the document]
Comment (kimkyung-goog) [reply google-cloud-clients/google-cloud-pubsub/src/main/java/com/google/cloud/pubsub/v1/Publisher.java]: Agreed. The problem is that it is difficult to set this retry settings differently for each group of messages. Maybe, we can have two publisherStub objects that have different retrySettings. Then, we can use one for messages without ordering keys and the other one for messages with ordering keys.

File: google-cloud-clients/google-cloud-pubsub/src/main/java/com/google/cloud/pubsub/v1/Publisher.java [hunk 4]
```diff
@@ -128,9 +137,31 @@ private Publisher(Builder builder) throws IOException {
...
+          .setMaxAttempts(Integer.MAX_VALUE)
+          .setTotalTimeout(Duration.ofNanos(Long.MAX_VALUE));
+    }
+
+    Set<StatusCode.Code> retryCodes;
+    if (enableMessageOrdering) {
+      retryCodes = EnumSet.allOf(StatusCode.Code.class);
```
[the remark this answers is not in the document]
Comment (kimkyung-goog) [reply google-cloud-clients/google-cloud-pubsub/src/main/java/com/google/cloud/pubsub/v1/Publisher.java]: Sure. Let me work on it. Should I create an issue on this branch or master?
(1) Non-retryable errors should fail immediately.
(2) Publishes to the failed ordering keys should be rejected.
(3) The Publisher should provide Reset (or ResumePublish) method that resumes publishing.

File: google-cloud-clients/google-cloud-pubsub/src/main/java/com/google/cloud/pubsub/v1/Publisher.java [hunk 5]
```diff
@@ -242,29 +273,18 @@ public String getTopicNameString() {
...
       logger.log(Level.FINER, "Scheduling a batch for immediate sending.");
-      final OutstandingBatch finalBatchToSend = batchToSend;
-      executor.execute(
-          new Runnable() {
-            @Override
-            public void run() {
-              publishOutstandingBatch(finalBatchToSend);
-            }
-          });
+      publishAllOutstanding();
```
[the remark this answers is not in the document]
Comment (kimkyung-goog) [reply google-cloud-clients/google-cloud-pubsub/src/main/java/com/google/cloud/pubsub/v1/Publisher.java]: It is intended by the design ("Flushing Batches" section in the guideline doc). The idea is that we also want to send non-frequently updated keys quickly. Imagine that there is a key that is only updated once in a while. If we do not call publishAllOutstanding() here, the messages with the key will be sent only when the alarm is fired, which makes the avg latency for that key is always high.

File: google-cloud-clients/google-cloud-pubsub/src/main/java/com/google/cloud/pubsub/v1/Publisher.java [hunk 6]
```diff
@@ -204,30 +234,38 @@ public String getTopicNameString() {
...
-          batchToSend = messagesBatch.popOutstandingBatch();
+        if (batch.getMessagesCount() == getBatchingSettings().getElementCountThreshold()) {
+          batchToSend = batch.popOutstandingBatch();
         }
       }
+
       // Setup the next duration based delivery alarm if there are messages batched.
-      if (!messagesBatch.isEmpty()) {
+      if (!batch.isEmpty()) {
         setupDurationBasedPublishAlarm();
```
[the remark this answers is not in the document]
Comment (kimkyung-goog) [reply google-cloud-clients/google-cloud-pubsub/src/main/java/com/google/cloud/pubsub/v1/Publisher.java]: That also makes sense, but I am concerned that we may end up having too many alarms. Recall that we expect hundreds of thousands of ordering keys.

File: google-cloud-clients/google-cloud-pubsub/src/main/java/com/google/cloud/pubsub/v1/Publisher.java (hunk 2 above)
Comment (kimkyung-goog) [reply google-cloud-clients/google-cloud-pubsub/src/main/java/com/google/cloud/pubsub/v1/Publisher.java]: Great idea.

File: google-cloud-clients/google-cloud-pubsub/src/main/java/com/google/cloud/pubsub/v1/SequentialExecutorService.java [hunk 7]
```diff
@@ -0,0 +1,263 @@
...
+interface CancellableRunnable extends Runnable {
+  public void cancel(Throwable e);
+}
+
+/**
+ * An executor service that runs the tasks with the same key sequentially. The tasks with the same
+ * key will be run only when its predecessor has been completed while tasks with different keys can
+ * be run in parallel.
+ */
+final class SequentialExecutorService<T> {
```
[the remark this answers is not in the document]
Comment (kimkyung-goog) [reply google-cloud-clients/google-cloud-pubsub/src/main/java/com/google/cloud/pubsub/v1/SequentialExecutorService.java]: I would love to it. Before merging (into master) or after?

File: google-cloud-clients/google-cloud-pubsub/src/test/java/com/google/cloud/pubsub/v1/FakePublisherServiceImpl.java [hunk 8]
```diff
@@ -33,6 +34,8 @@
 
   private final LinkedBlockingQueue<PublishRequest> requests = new LinkedBlockingQueue<>();
   private final LinkedBlockingQueue<Response> publishResponses = new LinkedBlockingQueue<>();
+  private final AtomicInteger nextMessageId = new AtomicInteger(1);
+  private boolean autoPublishResponse;
```
[the remark this answers is not in the document]
Comment (kimkyung-goog) [reply google-cloud-clients/google-cloud-pubsub/src/test/java/com/google/cloud/pubsub/v1/FakePublisherServiceImpl.java]: That makes sense.

File: google-cloud-clients/google-cloud-pubsub/src/main/java/com/google/cloud/pubsub/v1/SequentialExecutorService.java (hunk 7 above)
[the remark this answers is not in the document]
Comment (sduskis) [reply google-cloud-clients/google-cloud-pubsub/src/main/java/com/google/cloud/pubsub/v1/SequentialExecutorService.java]: Let's do it before merging to master.

File: google-cloud-clients/google-cloud-pubsub/src/main/java/com/google/cloud/pubsub/v1/Publisher.java (hunk 4 above)
[the remark this answers is not in the document]
Comment (sduskis) [reply google-cloud-clients/google-cloud-pubsub/src/main/java/com/google/cloud/pubsub/v1/Publisher.java]: The issues are all on master.  A `TODO` might be better here.

**Pull request closed without merging**